jobfunnel/tools/filters.py

```diff
@@ -1,59 +1,124 @@
+import nltk
+import logging
+
 from sklearn.feature_extraction.text import TfidfVectorizer
 from sklearn.metrics.pairwise import cosine_similarity
-from sklearn.neighbors import NearestNeighbors
-from typing import List, Dict
-import os
-import numpy as np
-import pickle
-import string
-import logging
+from typing import Dict, Optional
+from numpy import delete as np_delete, max as np_max, fill_diagonal
+
+
+def id_filter(cur_dict: Dict[str, dict], prev_dict: Dict[str, dict], provider):
+    """ Filter duplicates on job-id per provider
+        Args:
+            cur_dict: today's job scrape dict
+            prev_dict: the existing master list job dict
+            provider: job board used
+
+    """
+    # Get job_ids from scrape and master list by provider as lists
+    cur_job_ids = [job['id'] for job in cur_dict.values()]
+    prev_job_ids = [job['id'] for job in prev_dict.values()
+                    if job['provider'] == provider]
+
+    # Pop duplicate job ids from current scrape
+    duplicate_ids = [cur_dict.pop(job_id)['id'] for job_id in cur_job_ids
+                     if job_id in prev_job_ids]
+
+    # log duplicate id's
+    logging.info(f'found {len(cur_dict.keys())} unique job ids and '
+                 f'{len(duplicate_ids)} duplicates from {provider}')
+
 
-def tfidf_filter(cur_dict: Dict[str, dict], prev_dict: Dict[str, dict],
+def tfidf_filter(cur_dict: Dict[str, dict],
+                 prev_dict: Optional[Dict[str, dict]] = None,
                  max_similarity: float = 0.75):
     """ Fit a TFIDF vectorizer to a corpus of all listing's text
 
         Args:
-            cur_dict: the existing masterlist job dict
-            prev_dict: today's job scrape dict
-            max_similarity: threshold above which a blurb similarity = duplicate
+            cur_dict: today's job scrape dict
+            prev_dict: the existing master list job dict
+            max_similarity: threshold above which blurb similarity = duplicate
 
         Returns:
             list of duplicate job ids which were removed from cur_dict
-
-        @TODO skip calculating metric for jobs which have the same job id!
     """
+    # Retrieve stopwords if not already downloaded.
+    try:
+        stopwords = nltk.corpus.stopwords.words('english')
+    except LookupError:
+        try:
+            nltk.download('stopwords', quiet=True)
```

Comment:
Need to assign stopwords after downloading here as well
`stopwords = nltk.corpus.stopwords.words('english')`

Reply:
Good catch, I just put a fix out for it.